Assets/_Scripts/CurveController.cs

```diff
@@ -1,96 +1,138 @@
 using System.Collections.Generic;
 using UnityEngine;
-using UnityEngine.Splines;
 
 public class CurveController : MonoBehaviour
 {
     [SerializeField]
-    GameObject marker;
+    private GameObject marker;
 
-    [SerializeField]
-    InputManager inputManager;
-
-    [SerializeField]
-    MarkerSelection markerSelection;
-
-    [SerializeField]
-    SplineController splineController;
+    //[SerializeField] private InputManager inputManager;
 
     [SerializeField]
-    GhostKnotController ghostKnotController;
+    private GhostKnotController ghostKnotController;
 
     [SerializeField]
-    GameObject UI_FrameInputWindow;
+    private GameObject UI_FrameInputWindow;
 
-    private bool leftMouseButtonIsPressed = false;
-    private GameObject selectedMarker;
-    private GameObject selectedObject;
     public List<GameObject> markerList = new List<GameObject>();
     public List<int> ghostIndices = new List<int>();
 
+    private void Start()
+    {
+        //=======   Registering Events.   =======
+        GameEventSystem.Instance.RegisterListener<onRightClickPerformed>(AddMarkerHandler);
+        GameEventSystem.Instance.RegisterListener<MoveSelectedEvent>(MoveSelectMarker);
+        GameEventSystem.Instance.RegisterListener<DeleteSelectedEvent>(DeleteMarker_performed);
+        GameEventSystem.Instance.RegisterListener<ShiftDeleteSelectedEvent>(
+            ShiftDeleteMarker_performed
+        );
+        GameEventSystem.Instance.RegisterListener<onRecordingFinishedEvent>(AddRecordedMarkers);
+    }
+
+    private void OnDestroy()
+    {
+        //=======   Unregistering Events.   =======
+        GameEventSystem.Instance.UnregisterListener<onRightClickPerformed>(AddMarkerHandler);
+        GameEventSystem.Instance.UnregisterListener<MoveSelectedEvent>(MoveSelectMarker);
+        GameEventSystem.Instance.RegisterListener<DeleteSelectedEvent>(DeleteMarker_performed);
+        GameEventSystem.Instance.RegisterListener<ShiftDeleteSelectedEvent>(
+            ShiftDeleteMarker_performed
+        );
+        GameEventSystem.Instance.RegisterListener<onRecordingFinishedEvent>(AddRecordedMarkers);
+    }
+
     private void Update()
     {
-        if (markerSelection.selectedMarkerList != null && leftMouseButtonIsPressed)
-        {
-            splineController.UpdateSpline(markerList);
-            UpdateGhostMarkers();
-            //Update rotation of markers.
-            UpdateALLMarkersRotation();
-        }
+        //TODO What was this meant to be anyway?
+        GameEventSystem.Instance.Raise<UpdateSplineEvent, bool>(new UpdateSplineEvent(markerList));
```

Comment:
We only need to update the spline when something has changed in the spline. So no need to raise this event in the update. Instead, you should move this to the end of the methods that change something to the markers. I think we don't have any event such as "OnMarkerMove" in this class. We can create one and subscribe to a move-input event.

Reply:
Actually in the previous version you were calling the UpdateAllMarkersRotation only when some markers were selected and the mouse click was pressed. :D